File: examples/triage/fileinfo.cpp
#include "fileinfo.h"
#include "entropy.h"
#include "fontsettings.h"
#include "theme.h"
#include "copyablelabel.h"
#include <QClipboard>
#include <QApplication>
#include <QToolTip>
#include <QPainter>
#include <QFontMetrics>
#include <QScrollArea>
#include <QtConcurrent/QtConcurrent>
#include <QFuture>
#include <QFutureWatcher>

void FileInfoWidget::addCopyableField(const QString& name, const QVariant& value)
{
	auto& [row, column] = this->m_fieldPosition;

	const auto valueLabel = new CopyableLabel(value.toString(), getThemeColor(AlphanumericHighlightColor));
	valueLabel->setFont(getMonospaceFont(this));

	this->m_layout->addWidget(new QLabel(name), row, column);
	this->m_layout->addWidget(valueLabel, row++, column + 1);
}

void FileInfoWidget::addCopyableFieldWithElide(const QString& name, const QVariant& value, int maxWidth)
{
	auto& [row, column] = this->m_fieldPosition;

	const auto fullText = value.toString();
	const auto font = getMonospaceFont(this);
	const QFontMetrics fontMetrics(font);
	auto elidedText = fontMetrics.elidedText(fullText, Qt::ElideMiddle, maxWidth);
	elidedText.replace(QChar(0x2026), "...");

	const auto valueLabel = new CopyableLabel(elidedText, getThemeColor(AlphanumericHighlightColor));
	valueLabel->setFont(font);
	valueLabel->setCopyText(fullText);
	valueLabel->setToolTip(fullText + "\n\nClick to Copy");
	valueLabel->setSizePolicy(QSizePolicy::Preferred, QSizePolicy::Preferred);
	valueLabel->setWordWrap(false);

	this->m_layout->addWidget(new QLabel(name), row, column);
	this->m_layout->addWidget(valueLabel, row++, column + 1);
}

void FileInfoWidget::addField(const QString& name, const QVariant& value)
{
	auto& [row, column] = this->m_fieldPosition;

	const auto valueLabel = new QLabel(value.toString());
	valueLabel->setFont(getMonospaceFont(this));

	this->m_layout->addWidget(new QLabel(name), row, column);
	this->m_layout->addWidget(valueLabel, row++, column + 1);
}

void FileInfoWidget::addHashFields(BinaryViewRef view)
{
	auto& [row, column] = this->m_fieldPosition;

	const auto hashFieldColor = getThemeColor(AlphanumericHighlightColor);
	auto md5Label = new CopyableLabel("Calculating...", hashFieldColor);
	auto sha1Label = new CopyableLabel("Calculating...", hashFieldColor);
	auto sha256Label = new CopyableLabel("Calculating...", hashFieldColor);
	md5Label->setFont(getMonospaceFont(this));
	sha1Label->setFont(getMonospaceFont(this));
	sha256Label->setFont(getMonospaceFont(this));

	this->m_layout->addWidget(new QLabel("MD5: "), row, column);
	this->m_layout->addWidget(md5Label, row++, column + 1);
	this->m_layout->addWidget(new QLabel("SHA-1: "), row, column);
	this->m_layout->addWidget(sha1Label, row++, column + 1);
	this->m_layout->addWidget(new QLabel("SHA-256: "), row, column);
	this->m_layout->addWidget(sha256Label, row++, column + 1);

	// Process the hash calculations in a separate thread and update the labels when done
	QPointer<QFutureWatcher<QStringList>> watcher = new QFutureWatcher<QStringList>(this);
	connect(watcher, &QFutureWatcher<QStringList>::finished, this, [watcher, md5Label, sha1Label, sha256Label]() {
		if (watcher)
		{
			if (const auto results = watcher->result(); results.size() == 3)
			{
				md5Label->setText(results[0]);
				sha1Label->setText(results[1]);
				sha256Label->setText(results[2]);
			}
			watcher->deleteLater();
		}
	});

	uint64_t totalSize = view->GetLength();
	QFuture<QStringList> future = QtConcurrent::run([view, totalSize, watcher]() {
		QCryptographicHash md5(QCryptographicHash::Md5);
		QCryptographicHash sha1(QCryptographicHash::Sha1);
		QCryptographicHash sha256(QCryptographicHash::Sha256);

		uint64_t offset = 0;
		const uint64_t maxChunkSize = 128 * 1024 * 1024;
		auto chunkBuffer = std::make_unique<char[]>(maxChunkSize);
		while (offset < totalSize)
		{
			if (watcher->isCanceled())
				return QStringList();

			const auto remainingBytes = totalSize - offset;
			const auto currentChunkSize = std::min(maxChunkSize, remainingBytes);
			const auto bytesRead = view->Read(chunkBuffer.get(), offset, currentChunkSize);
			if (bytesRead != currentChunkSize)
				return QStringList{"Error", "Error", "Error"};

			const auto chunkView = QByteArrayView(chunkBuffer.get(), currentChunkSize);
			md5.addData(chunkView);
			sha1.addData(chunkView);
			sha256.addData(chunkView);
			offset += currentChunkSize;
		}

		return QStringList {md5.result().toHex(), sha1.result().toHex(), sha256.result().toHex()};
	});

	watcher->setFuture(future);
	connect(this, &QObject::destroyed, this, [watcher]() {
		if (watcher && watcher->isRunning()) {
			watcher->cancel();
			watcher->waitForFinished();
		}
	});
}

FileInfoWidget::FileInfoWidget(QWidget* parent, BinaryViewRef bv, EntropyWidget* entropyWidget)
{
	this->m_layout = new QGridLayout();
	this->m_layout->setContentsMargins(0, 0, 0, 0);
	this->m_layout->setVerticalSpacing(1);

	const auto view = bv->GetParentView() ? bv->GetParentView() : bv;

	const auto file = bv->GetFile();
	const auto filePath = file->GetFilename();

	const QFontMetrics monoMetrics(getMonospaceFont(this));
	const int maxPathWidth = monoMetrics.horizontalAdvance(QString(64, '0'));

	this->addCopyableFieldWithElide("Path on disk: ", filePath.c_str(), maxPathWidth);

	// If triage view is opened from a project, show both actual filepath and path relative to project
	if (const auto fileProjectRef = file->GetProjectFile())
	{
		const auto projectFilePath = file->GetProjectFile()->GetPathInProject();
		this->addCopyableFieldWithElide("Path in project: ", projectFilePath.c_str(), maxPathWidth);
	}

	const auto fileSize = QString::number(view->GetLength(), 16).prepend("0x");
	this->addCopyableField("Size: ", fileSize);

	// Display entropy value from the entropy widget
	if (entropyWidget)
	{
		auto& [row, column] = this->m_fieldPosition;
		m_entropyLabel = new CopyableLabel("Calculating...", getThemeColor(AlphanumericHighlightColor));
		m_entropyLabel->setFont(getMonospaceFont(this));
		this->m_layout->addWidget(new QLabel("Entropy: "), row, column);
		this->m_layout->addWidget(m_entropyLabel, row++, column + 1);

		// Connect to entropy updates
		connect(entropyWidget, &EntropyWidget::entropyUpdated, this, &FileInfoWidget::updateEntropy);
	}

	this->addHashFields(view);

	const auto scaledWidth = UIContext::getScaledWindowSize(20, 20).width();
	this->m_layout->setColumnMinimumWidth(FileInfoWidget::m_maxColumns * 3 - 1, scaledWidth);
	this->m_layout->setColumnStretch(FileInfoWidget::m_maxColumns * 3 - 1, 1);
	setLayout(this->m_layout);
}


void FileInfoWidget::updateEntropy(double avgEntropy)
{
	if (m_entropyLabel)
	{
		const auto entropyStr = QString::number(avgEntropy, 'f', 6);
		m_entropyLabel->setText(entropyStr);
	}
}

File: python/examples/triage/file_info.py
import os
import hashlib
import threading
import binaryninjaui
from binaryninjaui import UIContext
from PySide6.QtCore import Qt
from PySide6.QtWidgets import QWidget, QLabel, QGridLayout
from binaryninja.enums import ThemeColor


class CopyableLabel(QLabel):
	def __init__(self, text, color):
		super(CopyableLabel, self).__init__(text)
		from PySide6.QtGui import QPalette
		style = QPalette(self.palette())
		style.setColor(QPalette.WindowText, color)
		self.setPalette(style)
		self.setFont(binaryninjaui.getMonospaceFont(self))
		self.setCursor(Qt.PointingHandCursor)
		self.setMouseTracking(True)
		self.setToolTip("Click to Copy")
		self.copy_text = ""  # Empty - will fall back to current text

	def mousePressEvent(self, event):
		if event.button() == Qt.LeftButton:
			from PySide6.QtWidgets import QApplication
			# Fallback to current text if copy_text not set
			if self.copy_text:
				QApplication.clipboard().setText(self.copy_text)
			else:
				QApplication.clipboard().setText(self.text())

	def setCopyText(self, text):
		self.copy_text = text


class FileInfoWidget(QWidget):
	def __init__(self, parent, data, entropy_widget):
		super(FileInfoWidget, self).__init__(parent)
		self.data = data
		self.entropy_widget = entropy_widget

		layout = QGridLayout()
		layout.setContentsMargins(0, 0, 0, 0)
		layout.setVerticalSpacing(1)

		row = 0

		# Get the file path
		file_path = data.file.original_filename
		view = data.parent_view if data.parent_view else data

		# Add path on disk
		self._add_field(layout, row, "Path on disk:", file_path)
		row += 1

		# Add path in project if available
		if data.file.project_file:
			project_path = data.file.project_file.path_in_project
			self._add_field(layout, row, "Path in project:", project_path)
			row += 1

		# Add file size
		file_size = f"0x{view.length:x}"
		self._add_copyable_field(layout, row, "Size:", file_size)
		row += 1

		# Add entropy (will be updated when calculation completes)
		self.entropy_label = CopyableLabel("Calculating...",
			binaryninjaui.getThemeColor(ThemeColor.AlphanumericHighlightColor))
		layout.addWidget(QLabel("Entropy:"), row, 0)
		layout.addWidget(self.entropy_label, row, 1)
		row += 1

		# Add hash fields (calculated in background)
		self.md5_label = CopyableLabel("Calculating...",
			binaryninjaui.getThemeColor(ThemeColor.AlphanumericHighlightColor))
		layout.addWidget(QLabel("MD5:"), row, 0)
		layout.addWidget(self.md5_label, row, 1)
		row += 1

		self.sha1_label = CopyableLabel("Calculating...",
			binaryninjaui.getThemeColor(ThemeColor.AlphanumericHighlightColor))
		layout.addWidget(QLabel("SHA-1:"), row, 0)
		layout.addWidget(self.sha1_label, row, 1)
		row += 1

		self.sha256_label = CopyableLabel("Calculating...",
			binaryninjaui.getThemeColor(ThemeColor.AlphanumericHighlightColor))
		layout.addWidget(QLabel("SHA-256:"), row, 0)
		layout.addWidget(self.sha256_label, row, 1)
		row += 1

		# Set column stretch so value column expands
		layout.setColumnStretch(1, 1)

		self.setLayout(layout)

		# Connect to entropy widget signal
		if entropy_widget:
			entropy_widget.entropyUpdated.connect(self._update_entropy)

		# Start hash calculation in background thread
		self.hash_thread = threading.Thread(target=self._calculate_hashes)
		self.hash_thread.daemon = True
		self.hash_thread.start()

	def _add_field(self, layout, row, name, value):
		value_label = QLabel(value)
		value_label.setFont(binaryninjaui.getMonospaceFont(self))
		layout.addWidget(QLabel(name), row, 0)
		layout.addWidget(value_label, row, 1)

	def _add_copyable_field(self, layout, row, name, value):
		value_label = CopyableLabel(value,
			binaryninjaui.getThemeColor(ThemeColor.AlphanumericHighlightColor))
		layout.addWidget(QLabel(name), row, 0)
		layout.addWidget(value_label, row, 1)

	def _calculate_hashes(self):
		view = self.data.parent_view if self.data.parent_view else self.data
		total_size = view.length

		md5 = hashlib.md5()
		sha1 = hashlib.sha1()
		sha256 = hashlib.sha256()

		offset = 0
		chunk_size = 128 * 1024 * 1024  # 128MB chunks

		while offset < total_size:
			remaining = total_size - offset
			current_chunk = min(chunk_size, remaining)
			data = view.read(offset, current_chunk)

			if data is None or len(data) != current_chunk:
				self.md5_label.setText("Error")
				self.sha1_label.setText("Error")
				self.sha256_label.setText("Error")
				return

			md5.update(data)
			sha1.update(data)
			sha256.update(data)
			offset += current_chunk

		# Update labels on main thread
		from PySide6.QtCore import QMetaObject, Q_ARG
		QMetaObject.invokeMethod(self.md5_label, "setText", Qt.QueuedConnection, Q_ARG(str, md5.hexdigest()))
		QMetaObject.invokeMethod(self.sha1_label, "setText", Qt.QueuedConnection, Q_ARG(str, sha1.hexdigest()))
		QMetaObject.invokeMethod(self.sha256_label, "setText", Qt.QueuedConnection, Q_ARG(str, sha256.hexdigest()))

	def _update_entropy(self, avg_entropy):
		self.entropy_label.setText(f"{avg_entropy:.6f}")
